Structured logging

bbguimaraes opened the pull request:
@stevekuznetsov 

Some examples of the output:

https://github.com/bbguimaraes/origin-ci-tool/tree/structured-logging-ex

bbguimaraes commented on oct/ansible/oct/callback_plugins/log_results.py:
```diff
@@ -1,16 +1,44 @@
+    wrapper_class=structlog.stdlib.BoundLogger,
+    cache_logger_on_first_use=True)
+handler = logging.FileHandler(
+    join(os.environ.get('ANSIBLE_LOG_ROOT_PATH'), 'log.txt'), 'a')
```
Is this used somewhere outside this tool?  Could we change it to a single file?

stevekuznetsov replied:
Unfortunately callback plugins need to be self-contained.

bbguimaraes replied:
With the code as it is on this PR, the log file would be at

    ~/.config/origin-ci-tool/logs/log.txt

Can we change it to a single log file, e.g.

    ~/.config/origin-ci-tool/log.txt

or would it impact other places that look for the logs in this location?

There doesn't seem to be a reference to this path on this repository, but I
don't know about other systems.

stevekuznetsov replied:
Single log file sounds good to me, but I think we still need that  environment variable to point to the containing directory. We set it in `oct/config/ansible_client.py` to communicate to the plugins where to put logs.

bbguimaraes commented on oct/ansible/oct/callback_plugins/log_results.py:
```diff
@@ -380,24 +350,7 @@ def v2_runner_on_async_failed(self, result):
-        self.write_log('No hosts remaining.')
+        logger.error()
 
         # TODO: unimplemented methods follow - do we need them?
```
After a year without any change, can we answer "no" to this?

stevekuznetsov replied:
Today let's just put a `logger.info()` in these and make sure we implement every upstream callback.

stevekuznetsov commented on oct/ansible/oct/callback_plugins/log_results.py:
```diff
@@ -1,16 +1,44 @@
-from shutil import rmtree
-from tempfile import mkdtemp
-from traceback import format_exc
+import contextlib
```
In the project we always prefer `from X import Y` style if possible

stevekuznetsov commented on oct/ansible/oct/callback_plugins/log_results.py:
```diff
@@ -54,182 +79,105 @@ def log_exceptions_v2_playbook_on_start(func):
+    twice.
+    """
+    for key in data.keys():
+        if key.startswith('_ansible'):
```
Looks like we may need to make this a recursive sanitize:

```json
{
  "callback": "v2_runner_on_skipped",
  "host": "localhost",
  "result": {
    "msg": "All items completed",
    "changed": false,
    "results": [
      {
        "skipped": true,
        "_ansible_no_log": false,
        "skip_reason": "Conditional check failed",
        "_ansible_item_result": true,
        "item": "origin_ci_inventory_dir",
        "changed": false
},
```

stevekuznetsov commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(3) [CI] [oct] Implement the event stream logger for the Origin CI Tool](https://trello.com/c/RuVMEjin/128-3-ci-oct-implement-the-event-stream-logger-for-the-origin-ci-tool)

bbguimaraes commented:
@stevekuznetsov: updated the code and the examples.

Unfortunately my response to your first comment is now hidden, but please take
a look at it.  I noticed I need to add code to create the directory (if we
decide to keep it a directory), but I'll wait for your response.

stevekuznetsov commented on oct/ansible/oct/callback_plugins/log_results.py:
```diff
@@ -150,13 +150,19 @@ def sanitize_results(data):
-        for result in data['results']:
-            if 'stdout' in result and 'stdout_lines' in result:
-                del result['stdout_lines']
+    if isinstance(data, list):
```
Can we have nested `results` fields? If not, we should have a separate recursive function call for sanitizing just the `_ansible` bits -- a smell here is that the first call to `santiize_results()` will not ever be with a list, so we have two separate cases I think -- top-level where you want to get rid of `stdout_lines` and recursively where you want to get rid of keys

bbguimaraes commented:
@stevekuznetsov: updated.

There are other parts of the code that use the `logs` directory
(`generate_junit.py`, maybe others), so I kept the log file at the same place
and added`some code to the callback module's `__init__` to create the
directory if needed.

If there are no other comments, I think this is ready.

stevekuznetsov commented on oct/ansible/oct/callback_plugins/log_results.py:
```diff
@@ -55,181 +75,125 @@ def log_exceptions_v2_playbook_on_start(func):
+    """ Add the function name to the log. """
 
-    def __init__(self):
-        super(CallbackModule, self).__init__()
```
Why did we lose this call in the new `__init__`?

bbguimaraes replied:
Completely unintentional, thanks for catching that.  Fixed.

stevekuznetsov commented:
This looks really great. Lets get it in action!

stevekuznetsov commented:
Thanks @bbguimaraes
